CI: Add 48hr hold test

yadij opened the pull request.

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,52 @@
+  # (e.g., to skip a particular test until a known bug is fixed) or generate
+  # diffs against them (e.g., for `git diff --check`). This arbitrary limit
+  # tries to balance the two concerns.
+  CHECKOUT_FETCH_DEPTH: 1001
```
do we need to care for this any longer? In the end, it's github's resources, and they have plenty

yadij replied:
IMO the running time for the test is more what we care about. So avoiding a lot of data copy is likely to still be of benefit that way.

I am not sure what relation this value has to the "fetch-depth:" parameter on the job itself (one place where it was used). So left this matching the other files and set the job fetch-depth to the one commit we actually use.

kinkie replied:
Feels like premature optimziation. We're talking about seconds at worst. I won't block on this, but I also don't think it should be included.

yadij replied:
Why not include?

yadij replied:
Okay, removed in c1d68a6

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,52 @@
+jobs:
+
+  hold-for-review:
+    # Alex Rousskov has been granted a 48hr delay
+    # to allow him a chance to veto new additions
+    # regardless of other Squid Project member votes.
```
Wouldn't everyone in the approver list get the same privilege? I don't see anything special in the code about Alex

yadij replied:
Well, he requested it and is the only one of us making use of it vs the older process of reverting changes.

yadij replied:
Reworded in 78ba5c7caf anyway.

rousskov reviewed (changes requested):
At an implementation level, I am against introducing a hard-coded 48-hour delay for unanimously approved PRs. At a design level, we should avoid splitting complex PR merge management tasks between Anubis and GitHub Actions.

As always, I am happy to collaborate on changing what invariants Anubis should enforce.

> PR description: Currently we rely on Anubis bot to enforce the review hold. The
bot has not been completely reliable at this task - at times it has merged new code within 48hr of commits to edited PRs.

This statement is false or misleading -- Anubis has reliably enforced the "review hold" that it was tasked with enforcing. We can discuss changes to how the corresponding hold should be _defined_, but we should not accuse Anubis of being unreliable in this context.

> And there has been no application of the limit to github UI buttons (allowing manual commit at any time).

PRs that still require approval _are_ reflected in GitHub UI via non-green "PR approval" check managed by Anubis. Adding _another_ non-green check would not change much in this context as long as both checks are in sync. And adding a new PR approval check that is not in sync with the existing "PR approval" check is a bad idea for other obvious reasons. Again, if desired, we can change the definition of "[approved PR](https://github.com/measurement-factory/anubis/blob/master/README.md#voting-and-pr-approvals)", of course. Please propose a new definition if you think these changes are warranted; we can discuss how to implement the new definition once we agree on what that new definition should be.

P.S. If the goal is to block core developers from merging PRs prematurely, then we can also enhance Anubis to block otherwise-all-green PRs until their staging checks are complete. We can enlarge the existing "PR approval" check to cover more than just PR approvals OR we can add another check dedicated to staging tests. IMO, it is better to trust core developers not to merge manually instead, but I do not insist on that simple/cheap strategy. I am happy to collaborate on related improvements.

yadij commented:
> At an implementation level, I am against introducing a hard-coded 48-hour delay for unanimously approved PRs. 

Ah, good catch. Now added overrides for that and for reviewer vetos.
Right now this check is failing because you have a "request changes".

We can further test it works correctly once you clear your review to approved, and if kinkie also approves within 48hrs of now we can trace the override for unanimous approval.


At a design level, we should avoid splitting complex PR merge management tasks between Anubis and GitHub Actions.
> 
> As always, I am happy to collaborate on changing what invariants Anubis should enforce.
> 
> > PR description: Currently we rely on Anubis bot to enforce the review hold. The
> > bot has not been completely reliable at this task - at times it has merged new code within 48hr of commits to edited PRs.
> 
> This statement is false or misleading -- Anubis has reliably enforced the "review hold" that it was tasked with enforcing. We can discuss changes to how the corresponding hold should be _defined_, but we should not accuse Anubis of being unreliable in this context.

Like I wrote. There have been times when it merged PRs early and dropped last minute changes the author had made.

FTR; The specific case seems to be:
1)  commit X is pushed to the PR
2) reviewers immediately approve changes in commits A..X
3) Anubis begins merge procedure using "auto" branch
4) additional commits Y..Z arrive on PR branch
5) Anubis completes merge procedure for commit A..X
6) Anubis closes the PR.

These cases were only spotted when I went back to clean up my local branches found that github was holding references to closed PR work. Luckily these Y..Z commits tend to be docs or style polish right at the end and did not matter. But the fact that they can occur at all is a worry.


> P.S. If the goal is to block core developers from merging PRs prematurely, then we can also enhance Anubis to block otherwise-all-green PRs until their staging checks are complete. We can enlarge the existing "PR approval" check to cover more than just PR approvals OR we can add another check dedicated to staging tests. IMO, it is better to trust core developers not to merge manually instead, but I do not insist on that simple/cheap strategy. I am happy to collaborate on related improvements.

The goal is to make github UI perform more inline with what the Squid Project procedure requires. That should be matching what Anubis does, though not always - as I have mentioned there are edge cases where Anubis does what **you** wanted to implement, not what the Project procedure documents as requirements.

rousskov reviewed (changes requested):
> There have been times when it merged PRs early and dropped last minute changes the author had made.
> 
> FTR; The specific case seems to be:
> 
>     1. commit X is pushed to the PR 
>     2. reviewers immediately approve changes in commits A..X 
>     3. Anubis begins merge procedure using "auto" branch 
>     4. additional commits Y..Z arrive on PR branch 
>     5. Anubis completes merge procedure for commit A..X 
>     6. Anubis closes the PR.

Anubis is supposed to (and has code to) abort merging process after event 4. I believe I have seen Anubis doing that. If you have a specific PR where that abort did not happen, please share that PR number with me. If confirmed, we will need to investigate and fix that bug! That failure may be too old to investigate now; please alert me if you spot this behavior again.

----

> The goal is to make github UI perform more inline with what the Squid Project procedure requires. That should be matching what Anubis does, though not always - as I have mentioned there are edge cases where Anubis does what **you** wanted to implement, not what the Project procedure documents as requirements.

I disagree that Anubis goes against Project requirements. We should discuss any specific cases and reach an agreement _before_ proceeding with this or similar PRs (or related Anubis changes)!

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,101 @@
+            grep CHANGES_REQUESTED && false
+
+      - name: Fast Track Approval
+        # Unanimous Approval from Squid Foundation review team overrides 48hr hold check.
```
```suggestion
        # Unanimous Approval from Squid Core team overrides 48hr hold check.
```

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,101 @@
+    runs-on: ubuntu-24.04
+
+    steps:
+      - name: Checkout Squid sources
```
We are missing one step: this file will be in all forks of the squid repository, and apply to them as well, which we do not want.
Please protect against this ([instructions](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/using-conditions-to-control-job-execution#example-only-run-job-for-specific-repository))

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,101 @@
+        run: |
+          test `gh pr view ${{ github.event.pull_request.number }} --json author,latestReviews \
+            --template '{{.author.login}} APPROVED {{range .latestReviews}}{{.author.login}} {{.state}} {{end}}' |
+            grep -E "(kinkie|rousskov|yadij) APPROVED" | wc -l` -eq 3
```
```suggestion
            grep -E "^(kinkie|rousskov|yadij) APPROVED" | wc -l` -eq 3
```
be more pedantic against unwanted approvals

yadij commented:
@rousskov AFAICS the issues you brought up about this PR change in earlier reviews have all been resolved. The remaining discussion is all about Anubis, which is not something this PR can alter.  Please re-review the proposed change for technical issues (adherence to the documented voting procedure).

@kinkie some updates have changed the code. please re-review.

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,105 @@
+        run: |
+          gh pr view ${{ github.event.pull_request.number }} --json latestReviews \
+            --template '{{range .latestReviews}}{{.state}}{{end}}' |
+            grep CHANGES_REQUESTED && false
```
```suggestion
            grep -q CHANGES_REQUESTED && false
```

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,105 @@
+
+on:
+  schedule:
+    - cron: "10 1 * * *" # once a day
```
Why not more often? Say every 4 hours or so?

yadij replied:
The criteria calculations have a 1-day resolution. So it is not worth doing more than once per day after the initial check.

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,105 @@
+            grep CHANGES_REQUESTED && false
+
+      - name: Fast Track Approval
+        # Unanimous Approval from Squid Core team overrides 48hr hold check.
```
3 from 4 is not unanimous, please correct

yadij replied:
done

kinkie commented on .github/workflows/review.yaml:
```diff
@@ -0,0 +1,105 @@
+        run: |
+          test `gh pr view ${{ github.event.pull_request.number }} --json author,latestReviews \
+            --template ' {{.author.login}} APPROVED {{range .latestReviews}}{{.author.login}} {{.state}} {{end}}' |
+            grep -o -E " (hno|kinkie|rousskov|yadij) APPROVED" | wc -l` -eq 2
```
Does authorship appear as APPROVED here?

kinkie reviewed (changes requested):
Small but meaningful change requests

yadij commented:
Clearing this up for merge since issues were all resolved months ago with no further feedback.

yadij commented:
> > [yadij](https://github.com/yadij) yadij dismissed stale reviews from [kinkie](https://github.com/kinkie) and [rousskov](https://github.com/rousskov) [9 hours ago](https://github.com/squid-cache/squid/pull/2067#event-21633894517)
> 
> @yadij, you do not have the right to dismiss negative votes of other core developers.

1) See https://github.com/squid-cache/squid/pull/2067#issuecomment-3238887339.  Per https://wiki.squid-cache.org/MergeProcedure#votes item 1, only blocks until resolved.

 2) After waiting 3 months for any type of response, I applied https://wiki.squid-cache.org/MergeProcedure#exceptions item 1 to clear the merge queue item.

 3) You applied https://wiki.squid-cache.org/MergeProcedure#exceptions item 2 to revert.

yadij commented:
Oh, and Anubis did not wait the required 48hr hold period between my change [c4ce4ee](https://github.com/squid-cache/squid/pull/2067/commits/c4ce4eef48729bf8ca790efd786e52e91306af2b) and starting the merge checks with auto branch.
Per Squid Project policy you are supposed to have 48hrs to re-review and/or clear my merge label.
